Add transpiled langauges support

Veetaha opened the pull request:
There is one inconvenience when using this package. The protocol of exporting the `Adapter` class as `module.exports` itself. I wish there was a possibility to change this from top-level export to some property on the export object e.g. `module.exports.Adapter`.
This is inconvenient because you cannot have both top-level and regular exports in the same file in typescript... E.g. this doesn't work
```ts
export type Client = elasticsearch.Client;

// Compile error: 
// An export assignment cannot be used in a module with other exported elements.ts(2309)
export = class ElasticsearchAdapter { /* ... */ }

```
The better inferface would be
```ts
export class Adapter { /* */ }
export type Client = /* anything else I'd like to export from this module */
```

Anyway, what this PR actually does, is it gives the TypeScript declarations for adapter interface:
![east](https://user-images.githubusercontent.com/36276403/78139466-3931b580-7431-11ea-962c-fb0fbbb3e907.gif)

okv commented:
Hi.
I have some thoughts about typescript support:

1) I don't use typescript much so correct me if I'm wrong, east actually doesn't export adapter or it's constructor, so it looks a bit strange to provide typescript declaration for those. And related question here is how are you using this particular types?

2) I think typescript support should appear in east not as separate feature but as reasonable set of features  e.g.:
* implement options discussed [here](https://github.com/okv/east/pull/48#issuecomment-595784669)
* discuss and provide types
* document typescript support at readme

Currently I'm ready to discuss such features, manage PR's but I don't have time to implement it.

Veetaha commented:
1. The types are used with [`import type { Adapter, AdpaterConstructor } from "east"` syntax](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#-type-only-imports-and-export). This means that in the transpiled JavaScript code this `import` will be entirely removed. The users just get the `interface`s that they should implement. `Adapter` and `AdapterConstructor` symbols can be used in type-only contexts. So you cannot e.g. write this:
```ts
import type { AdapterConstructor } from "east";

// Compile error:
// 'AdapterConstructor' only refers to a type, but is being used as a value here.ts(2693)
const abstractConstructor = new AdapterConstructor<unknown>();
```
The purpose of these two interfaces is to provide **(a)** type checks that the adapter the user has created does abide by the protocol defined by east (i.e. has all the required methods and methods have the appropriate parameter and return types (in other words - signature)). To achieve this the user explicitly adds `implements` clause to the adapter class:
```ts
import { DbClient } from "some-mainstream-database";
import type { Adapter, AdapterConstructor } from "east";

// Compile error:
// Class 'MyAdapter' incorrectly implements interface 'Adapter<DbClient>'.
// Type 'MyAdapter' is missing the following properties from type 'Adapter<DbClient>': disconnect, markExecuted ... (2420)
class MyAdapter implements Adapter<DbClient> {
    constructor(params: unknown) { /* */ }

    async connect(): Promise<DbClient> { /* */ }
    getTemplatePath(): string { /* */ }
    // Compile error: (the method returns Promise<number[]> instead of Promise<string[]>)
    // Type  'number' is not assignable to type 'string'.ts(2416)
    getExecutedMigrationNames(): Promise<number[]> { /* */ }
}

// At this line we check that  `MyAdapter` (i.e. the desugared constructor function) 
// conforms to `AdapterConstructor<DbClient>` type,
// i.e. that we can call `new MyAdapter()` and
// it returns `MyAdapter<DbClient>` as per the `AdapterConstructor` interface
const _: AdapterConstructor<DbClient> = MyAdapter;

export = MyAdapter;
```
**(b)** the interface declaration documents the protocol via the doc comments:
![image](https://user-images.githubusercontent.com/36276403/78376488-16410600-75d7-11ea-810c-3106a63be67a.png)

2. Hmm, I'll try to look into this.

Veetaha commented:
@okv I updated the PR with the following points done:
- [x] implement options discussed here
- [x] document typescript support at readme
- [ ] discuss and provide types

I didn't add type declarations for `MigrationManager`, but I think this will be straightforward. I'd like to postpone this untill you review the PR.

okv commented:
I see a lot of work done here, I'll try to take a look shortly

Veetaha commented on README.md:
```diff
@@ -144,14 +145,14 @@ exports.rollback = function(client, done) {
-* `client` is connect to current db and he determined by adapter (see [adapters](#adapters) section)
-* `done` is function which should be called at the end of migration (if any
-error occured you can pass it as first argument)
+* `client` is represents a connection to the current db and he is determined by the adapter (see [adapters](#adapters) section)
```
```suggestion
* `client` represents a connection to the current db and he is determined by the adapter (see [adapters](#adapters) section)
```

okv commented on testUtils/createEnv.js:
```diff
@@ -42,9 +42,34 @@ module.exports = (params) => {
-			return createEastrc({dir, configParams: {dir: migrationsDir}});
+			const eastrc = {};
+
+			if (params.migratorParams && params.migratorParams.configureParams) {
```
if I'm not mistaken all params except `dir` (which is required to put `.eastrc` in right place) could be passed via `params.migratorParams.configureParams` (which will be passed to `migrator.configure`).

Veetaha replied:
I removed `migration*Extension` setup from here and left only `dir` and `sourceDir` propagation to test migrator in https://github.com/okv/east/pull/54/commits/bc6b8ad847d5b3bd97ff2234dc259d6669b9a956

okv replied:
that's ok

okv commented on lib/migrator/methods/index.js:
```diff
@@ -16,12 +16,21 @@ exports.rollback = require('./rollback');
 
 // Other API methods
 exports.getAllMigrationNames = require('./getAllMigrationNames');
+exports.getAllMigrationSourceNames = require('./getAllMigrationSourceNames');
```
I think it would be better extend existing migrator method instead of adding new ones:

* getMigrationSourcePathByName(name) -> getMigrationPathByName(name, [migrationType])
* getMigrationSourceNameByPath(path) -> getMigrationNameByPath(path, [migrationType])
* getAllMigrationSourceNames() -> getAllMigrationNames([migrationType])
* isSourceDirExists() -> isDirExists([migrationType])
* isMigrationSourceExists(name) -> isMigrationExists([migrationType])

For all method `migrationType` is "executable" by default but could be set to "source". Internally such type could be resolved by some internal helper e.g. `_getMigrationTypeParams(migrationType)` which returns object with appropriate `{dir, extension}` or throws an error for unrecognized type.

In my opinion it also looks good for migration manager:

* manager.getMigrationSourcePath(name) -> getMigrationPath(name, [migrationType])

For me it looks like job is done with less effort and migrator, manager api are still good. What do you think?

Veetaha replied:
Yeah, think it is reasonable to make an enum `"executable" | "source"`.

Veetaha replied:
Done in https://github.com/okv/east/pull/54/commits/69649ef8b8694eb82d751a78cddbddea7761abc1

Veetaha commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ node_modules/
 .nyc_output
 coverage
 migrations/
+.vscode
```
Not in my case, I also have a language server project which includes writing a custom vscode extension and we do commit the `.vscode` dir with configurations for debugging. This may look like an exceptional case, so I'd rather remove this change and just not commit the `.vscode` dir.

Veetaha replied:
Done in https://github.com/okv/east/pull/54/commits/77a3c0fd31e73eb2b839d0485e0511ca5ac73bcf

Veetaha commented on bin/east:
```diff
@@ -39,9 +46,17 @@ program
 			.then((migrationParams) => {
+				const {dir, sourceDir} = migrationParams;
+
+				let messageSuffix = `${'`'}${dir}${'`'}`;
```
I wonder why I didn't know that you can use backslash escapes in a template literal...

Veetaha replied:
Done in https://github.com/okv/east/pull/54/commits/6eefe043b806dae9b3f9b2a8769b2ab648ed3765

Veetaha commented on lib/migrator/methods/configure.js:
```diff
@@ -10,7 +10,13 @@ module.exports = function configure(params) {
 			this.params = {
 				// migrations dir
-				dir: pathUtils.resolve('migrations'),
+				dir: null,
```
This will mean having an odd setup like:
When the user specifies only `--migrationExtension ts`
then `--sourceMigrationExtension` is left to be `js` by default, and this is very improbable that the user compiles `js` to `ts`.

Maybe we should not only set `--sourceDir` equal to `dir` by default, but also `--sourceMigrationExtension`.

As for passing only `--sourceMigrationExtension ts` it does seem resonable that `--migrationExtension` is left to be `js` by default.

And one more setup when the user passes only `--sourceDir tsFilesDir` it is not expected that `--dir` will be equal to `migrations`, because you'd rather expect it to be in `build` or `out` directories where your transpiled javascript files reside. Though maybe just documenting this more thoroughly will be enough.

Veetaha replied:
Done in https://github.com/okv/east/pull/54/commits/bc6b8ad847d5b3bd97ff2234dc259d6669b9a956

okv replied:
Your current implementation (when `dir`, `migrationExtension` are static and `sourceDir`, `sourceMigrationExtension` are equal to `dir` and `migrationExtension` by default) looks good to me. I think it's ok to set `dir` manually (if you don't like default "migrations" dir) when you set `sourceDir`. I suggest to document options default values at east binary option description.

okv replied:
You already documented it, nice)

Veetaha commented on testUtils/cleanSnapshotData.js:
```diff
@@ -1,7 +1,7 @@
 
 const migrationsDirRegExp = new RegExp(
-	'/.+/migrations',
+	'/.+?/migrations',
```
It happens so for the message produced by `east init` when you pass different values for `--dir` and `--sourceDir`, then you get the following message:
```
east init --dir ./migrations --sourceDir ./migrationsSource
initialization successfully done, migration executable and source files will 
be stored at `./migrations` and `./migrationsSource` respectively
```
So the regexp eagerly matches the whole `stored at (/migrations` and `./migrations)Source` string (I emphasized the match with parens).

Veetaha commented on bin/east:
```diff
@@ -39,9 +46,17 @@ program
 				this.logger.info(
-					'initialization successfully done, migration files will be ' +
-					'stored at: `' + migrationParams.dir + '`'
+					`${'initialization successfully done, migration executable ' +
```
Done in https://github.com/okv/east/pull/54/commits/6eefe043b806dae9b3f9b2a8769b2ab648ed3765

Veetaha commented on lib/migrator/methods/getAllMigrationNames.js:
```diff
@@ -10,14 +10,19 @@ module.exports = function getAllMigrationNames() {
+					const {migrationExtension} = this.params;
+
+					return fileName.endsWith(`.${migrationExtension}`) &&
+						this._nameRegExp.test(fileName.slice(
```
Changed to `pathUtils.parse()` in https://github.com/okv/east/pull/54/commits/69649ef8b8694eb82d751a78cddbddea7761abc1

Veetaha commented:
@okv please see last 4 commit

Veetaha commented on tap-snapshots/test-02-snapshot-bin-east-commands-create-withUnsupportedFileExtension.js-TAP.test.js:
```diff
@@ -0,0 +1,11 @@
+ */
+'use strict'
+exports[`test/02-snapshot/bin/east/commands/create/withUnsupportedFileExtension.js TAP > output 1`] = `
+Adapter doesn't provide template '.bruh' source files
```
Don't know whether this is good to make this a snapshot, because different adapters may produce different error messages and it is advertised that people writing custom adapters can test them with the tests from this repo...

okv replied:
Currently only test/01-migrator tests intend to use with custom adapters so It's good to have this test.

okv commented on README.md:
```diff
@@ -356,11 +356,9 @@ Returns *Promise<Boolean>*.
-
-* **getMigrationSourcePath(name)** - returns an absolute path of the migration source
-file on disk by the name of the migration. Returns *Promise<String>*.
+* **getMigrationPath(name, migrationFileType)** - returns an absolute path of the migration
```
Other migration manager methods also will be affected:

* `isInitialized` should check for existing of both dirs
* `getMigrationPath` should accept `migrationFileType`

okv commented on lib/migrator/methods/_getMigrationTypeParams.js:
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+module.exports =
+	function _getMigrationTypeParams(migrationFileType = 'executable') {
```
Maybe it should be called `_getMigrationFileTypeParams`? since it accepts `migrationFileType`

okv replied:
I think here `migrationFileType` should not have default value

okv commented on README.md:
```diff
@@ -71,24 +70,26 @@ Run `east -h` to see all commands:
-    --dir <dir>          dir where migrations stored
-    --url <url>          db connect url
-    --trace              verbose mode (including error stack trace)
+    -V, --version                       output the version number
```
Options should be updated, default values are documented now

okv commented:
> @okv please see last 4 commits
Done, it seems to be good

okv commented:
Looks like we are about ready to go. Main issues:

* nodejs 4 support should be fixed, green builds are required)
* would be nice to have typescript types for migration manager

p.s. Travis builds statuses should appear at PR (I changed settings).

okv commented on README.md:
```diff
@@ -453,6 +456,55 @@ your adapter e.g.
 root of the cloned repository.
 
 
+## TypeScript and other transpiled languages support
```
Info about this feature also should be added to CHANGELOG.md to section:

```md
## [Unreleased][]

### Added

```

Veetaha replied:
Which version will it be?

okv replied:
It should be "Unreleased" section but there is not obvious changelog linter requirement related to this, so let me do that.

Veetaha commented:
Now everything is strongly-typed and documented:
![types](https://user-images.githubusercontent.com/36276403/78850797-c45a0f00-7a20-11ea-986d-f24016e6aad7.gif)

Veetaha commented:
I have removed node 4 from travis, it fails due to the array destructuring syntax used in tests as well as in `esm.js` dependency which I have no clue where it comes from and how to fix

okv commented on lib/migrationManager.js:
```diff
@@ -24,7 +24,8 @@ MigrationManager.prototype.init = function init() {
 
 MigrationManager.prototype.isInitialized = function isInitialized() {
-	return this.migrator.isDirExists();
+	return this.migrator.isDirExists('executable')
```
This is not gonna work because both results are promises.

okv commented:
> Now everything is strongly-typed and documented:
> ![types](https://user-images.githubusercontent.com/36276403/78850797-c45a0f00-7a20-11ea-986d-f24016e6aad7.gif)

:+1:

okv commented:
> I have removed node 4 from travis, it fails due to the array destructuring syntax used in tests as well as in `esm.js` dependency which I have no clue where it comes from and how to fix

I just can't let go support of node.js 4 =)
I know it may not be that easy to support it, so my suggestion is to rebase that branch on feature/v2 branch (here is th PR #50), but then this feature will be released in v2 along with other changes (some of them are breaking, see the PR #50).

Veetaha commented:
@okv
Fixed the `isInitialized()` and rebased the branch on top of #50. The PR now also has #50's branch as target branch, so you should be able merge it into your feature branch and then merge that branch into master for 2.0 release.

okv commented:
> @okv
> Fixed the `isInitialized()` and rebased the branch on top of #50. The PR now also has #50's branch as target branch, so you should be able merge it into your feature branch and then merge that branch into master for 2.0 release.

:+1: great job, I'm going to take a look on final diff and merge it.
Thanks for your contribution.
I'll post updates (about changes, published versions) to #50
